File: 97. Interleaving String/97. Interleaving String top-down.py
from functools import lru_cache


class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1)+len(s2) != len(s3):
            return False
        if len(s1) == 0 and len(s2) == 0 and len(s3) == 0:
            return True

        @lru_cache(None)
        def dp(i: int, j: int, k: int) -> bool:
            # i,j,k are index of s1,s2,s3 accordingly
            if k == len(s3):
                if i == len(s1) and j == len(s2):
                    return True
                return False
            if i < len(s1) and s1[i] == s3[k]:
                if dp(i+1, j, k+1):
                    return True
            if j < len(s2) and s2[j] == s3[k]:
                if dp(i, j+1, k+1):
                    return True
            if i == len(s1)-1 and j == len(s2)-1 and k == len(s3)-1:
                return True
            return False

        return dp(0, 0, 0)


s = Solution()
# s1 = "aabcc"
# s2 = "dbbca"
# s3 = "aadbbbaccc"
s1 = "aabcc"
s2 = "dbbca"
s3 = "aadbbcbcac"
print(s.isInterleave(s1, s2, s3))

File: 97. Interleaving String/97. Interleaving String top-down.java
class Solution {
    private String s1, s2, s3;
    private int[][][] memo;

    private boolean dp(int i, int j, int k) {
        if (memo[i][j][k] == -1) {
            if (k == s3.length()) {
                if (i == s1.length() && j == s2.length()) {
                    memo[i][j][k] = 1;
                    return true;
                }
                memo[i][j][k] = 0;
                return false;
            }
            if (i < s1.length() && s1.charAt(i) == s3.charAt(k)) {
                if (dp(i + 1, j, k + 1)) {
                    memo[i][j][k] = 1;
                }
            }
            if (j < s2.length() && s2.charAt(j) == s3.charAt(k)) {
                if (dp(i, j + 1, k + 1)) {
                    memo[i][j][k] = 1;
                }
            }
            if (memo[i][j][k] == -1) {
                memo[i][j][k] = 0;
            }
        }
        return memo[i][j][k] == 1 ? true : false;
    }

    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) {
            return false;
        }
        if (s1.length() == 0 && s2.length() == 0 && s3.length() == 0) {
            return true;
        }
        this.s1 = s1;
        this.s2 = s2;
        this.s3 = s3;
        this.memo = new int[s1.length() + 1][s2.length() + 1][s3.length() + 1];
        for (int[][] i : memo) {
            for (int[] j : i) {
                Arrays.fill(j, -1);
            }
        }
        return dp(0, 0, 0);
    }
}
